ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -165,42 +176,68 @@ implAddTxs :: forall m blk. (MonadSTM m, ApplyTx blk)
+    (vr, removed, unvalidated, mempoolSize) <- atomically $ do
+      IS{isTip = initialISTip} <- readTVar mpEnvStateVar
+
       -- First sync the state, which might remove some transactions
-      syncRes@ValidationResult { vrInvalid = removed } <- validateIS mpEnv
-      -- Then validate the new transactions
-      let ValidationResult
-            { vrBefore, vrValid, vrLastTicketNo
-            , vrNewValid = accepted
-            , vrInvalid  = rejected
-            } = validateNew syncRes
-      writeTVar mpEnvStateVar IS { isTxs          = vrValid
-                                 , isTip          = vrBefore
-                                 , isLastTicketNo = vrLastTicketNo
-                                 }
-      mempoolSize <- getMempoolSize mpEnv
-      return (removed, accepted, rejected, mempoolSize)
+      syncRes@ValidationResult
+        { vrBefore
+        , vrValid
+        , vrInvalid = removed
+        , vrLastTicketNo
+        } <- validateIS mpEnv
+
+      -- Determine whether the tip was updated after a call to 'validateIS'
+      -- @intricate: explain why
+      if initialISTip /= vrBefore
+      then do
+        -- The tip changed.
+        -- Because 'validateNew' can 'retry', we'll commit this STM
+        -- transaction here to ensure that we don't lose any of the changes
+        -- brought about by 'validateIS' thus far.
+        writeTVar mpEnvStateVar IS { isTxs          = vrValid
+                                   , isTip          = vrBefore
+                                   , isLastTicketNo = vrLastTicketNo
+                                   }
+        mempoolSize <- getMempoolSize mpEnv
+        pure (syncRes, removed, txs, mempoolSize)
+      else do
+        -- The tip was unchanged.
+        -- Therefore, we don't have to worry about losing any changes in the
+        -- event that we have to 'retry' this STM transaction. Continue by
+        -- validating the provided new transactions.
+        (vr, unvalidated) <- validateNew mpEnv txs syncRes
+        mempoolSize       <- getMempoolSize mpEnv
+        pure (vr, removed, unvalidated, mempoolSize)
+
+    let ValidationResult
+          { vrNewValid = accepted
+          , vrInvalid  = rejected
+          } = vr
 
-    traceBatch TraceMempoolRemoveTxs   mempoolSize (map fst removed)
     traceBatch TraceMempoolAddTxs      mempoolSize accepted
     traceBatch TraceMempoolRejectedTxs mempoolSize (map fst rejected)
+    unless (null removed) $
+      traceBatch TraceMempoolRemoveTxs mempoolSize (map fst removed)
 
-    return $ [(tx, Just err) | (tx, err) <- rejected] ++
-             zip accepted (repeat Nothing)
+    case unvalidated of
+      -- All of the provided transactions have been validated.
+      [] -> return $ [(tx, Just err) | (tx, err) <- rejected] ++
+                     zip accepted (repeat Nothing)
+
+      -- There are still transactions that remain which need to be validated.
+      _  -> implAddTxs mpEnv unvalidated
```

Comment:
How do transactions that were validated in this call end up in the end result?

Reply:
I've resolved this locally but haven't pushed it yet. Thanks!